Add flex_max

RMeli opened the pull request:
This PR adds `--flex_max`m a soft limit for the number of flexible residues; only the `--flex_max` closest residues to the ligand are retained.

RMeli commented:
From a first visual inspection it seems to work but more _tests are needed_. 

One thing I realised is that a flexible residues can be selected because of backbone atom near the ligand; it would be more beneficial to consider only side-chain atoms in the search of flexible residues, since they are the ones that actually move? I don't know if there is an easy way to do this...

dkoes commented:
Assume standard pdb atom names to identify backbone?

RMeli commented on gninasrc/lib/flexinfo.cpp:
```diff
@@ -102,6 +103,76 @@ void FlexInfo::sanitizeResidues(OpenBabel::OBMol& receptor) {
+      // Loop over residue atoms
+      for(const auto ares: res->GetAtoms()){
+
+        if(!isSideChain(ares->GetResidue()->GetAtomID(&(*ares)))) continue; // skip backbone
```
I'm not very happy with `ares->GetResidue()->GetAtomID(&(*ares)))`. Is there a better way to get the PDB atom names (stored in `OBResidue`) from an `OBAtom`? I might have overlooked something in the documentation, but I couldn't find a better way of achieving this.

RMeli replied:
This line of code is nastier than I previously thought by running only the docking; many `gnina` test fail with this line but work when it's commented out (together with the other similar line)... Any suggestion on how to do this properly?

dkoes replied:
The residue is where the atom names are.  There's no need for &(*ares) - can just do ares.

What is the failure, is ares sometimes null?

RMeli replied:
It was a segmentation fault that I could not elucidate with `valgrind` (I'll check if it's `null`). Here I changed it to `res->GetAtomID(ares)` since I'm looping over residues first and this solves the issue. 

However, where I'm looping over atoms and I still need to use `a->GetResidue()->GetAtomID(&(*a))` (which still causes problems), and using `a->GetResidue()->GetAtomID(a)` as you suggest gives the following problem:

```
/biggin/b195/lina3015/Documents/git/gnina/gninasrc/lib/flexinfo.cpp: In member function 'void FlexInfo::extractFlex(OpenBabel::OBMol&, OpenBabel::OBMol&, std::__cxx11::string&)':
/biggin/b195/lina3015/Documents/git/gnina/gninasrc/lib/flexinfo.cpp:192:49: error: no matching function for call to 'OpenBabel::OBResidue::GetAtomID(OpenBabel::OBMolAtomIter&)'
     if(!isSideChain(a->GetResidue()->GetAtomID(a))) continue; // skip backbone
                                                 ^
In file included from /usr/local/include/openbabel3/openbabel/atom.h:34:0,
                 from /biggin/b195/lina3015/Documents/git/gnina/gninasrc/lib/box.h:3,
                 from /biggin/b195/lina3015/Documents/git/gnina/gninasrc/lib/flexinfo.h:6,
                 from /biggin/b195/lina3015/Documents/git/gnina/gninasrc/lib/flexinfo.cpp:1:
/usr/local/include/openbabel3/openbabel/residue.h:127:20: note: candidate: std::__cxx11::string OpenBabel::OBResidue::GetAtomID(OpenBabel::OBAtom*) const
     std::string    GetAtomID(OBAtom *atom)        const;
                    ^~~~~~~~~
/usr/local/include/openbabel3/openbabel/residue.h:127:20: note:   no known conversion for argument 1 from 'OpenBabel::OBMolAtomIter' to 'OpenBabel::OBAtom*'
```

dkoes replied:
Oh right, if it is an iterator and not a pointer you do need that ugliness..

RMeli replied:
`a->GetResidue()` is indeed `NULL` some times; I'll investigate further, thanks for the pointer ([xkcd.com/138/](https://www.xkcd.com/138/))!

dkoes commented:
A test case that makes sure the right residues are selected that that flexible docking actually changes the side chain conformations would be appreciated.

RMeli commented:
Sure! Can I add such tests to `test_flexmin.py` (to be renamed `test_flex.py`) or do you want a separate test?

dkoes commented:
Whatever makes the most sense to you.

RMeli commented:
The flexible docking test takes ~30s to complete with one flexible residue and ~90s with two flexible residues. Is the one with two flexible residues still acceptable?

dkoes commented:
Try setting --num_mc_steps to something low like 50.  That will do less sampling, speeding things up, while still testing multiple residues.

RMeli commented:
Great suggestion, thanks!

RMeli commented:
Tests are now in a decent shape (testing both docking with and without `--flex_max`, and CNN-minimisation). `gninaflex` takes ~60s to run (5x docking and 2x minimisation). I guess it's fine, compared to the ~700s for `gninacheck`?

The PR is almost ready to review. The only thing I have left to do is to understand why the `NULL` pointer in `FlexInfo::extractFlex` affects tests that have nothing to do with flexible residues.

RMeli commented on gninasrc/lib/flexinfo.cpp:
```diff
@@ -115,9 +192,15 @@ void FlexInfo::extractFlex(OpenBabel::OBMol& receptor, OpenBabel::OBMol& rigid,
+    if(a->GetAtomicNum() == 1) continue; //heavy atoms only
+    if (a->GetResidue() != NULL){ // Check
+      // TODO: @RMeli figure out exactly then this happens 
+      if (!isSideChain(a->GetResidue()->GetAtomID(&(*a))))
```
This causes a segmentation fault for the tests where an `.xyz` file is used (such as `grid2gpu`). The reason is that `a->GetResidue()` returns a `NULL` pointer in that situation (as already pointed out by @dkoes ). I feel that `if (a->GetResidue() != NULL)` is an hacky solution, do you have alternative suggestions? 

I think skipping the call `FlexInfo::extractFlex` in `MolGetter::create_init_model` could be better, but I'm not sure I fully understand all the code branches of  such function. Any advice?

RMeli replied:
I left `if (a->GetResidue() != NULL)` as a guard, but added an if/else block in `MolGetter::create_init_model` to call `FlexInfo::extractFlex` only with PDB files. I feel this is a better workaround.

RMeli commented:
Thank you for pointing out the problem; it was a dirty solution. I changed the check on file extension to a check on `NumResidues()`, which seems to solve the problem more generally. I left the `if (a->GetResidue() != NULL)` guard, but I added an exception in case the information is missing (in order not to silently ignore the problem).
